lib/vrl/stdlib/src/insert.rs

```diff
@@ -0,0 +1,140 @@
+use lookup_lib::{LookupBuf, SegmentBuf};
+use vrl::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct Insert;
+
+impl Function for Insert {
+    fn identifier(&self) -> &'static str {
+        "insert"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "target",
+                kind: kind::ANY,
+                required: true,
+            },
+            Parameter {
+                keyword: "field",
+                kind: kind::BYTES | kind::ARRAY,
+                required: true,
+            },
+            Parameter {
+                keyword: "value",
+                kind: kind::ANY,
+                required: true,
+            },
+        ]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "insert existing field",
+                source: r#"insert!({"foo": "bar"}, "foo", "baz")"#,
+                result: Ok(r#"{ "foo": "baz" }"#),
+            },
+            Example {
+                title: "nested fields",
+                source: r#"insert!({}, ["foo", "bar"], "baz")"#,
+                result: Ok(r#"{ "foo": { "bar" : "baz" } }"#),
+            },
+            Example {
+                title: "indexing",
+                source: r#"insert!({"foo": { "bar": [] }}, ["foo", "bar", 1], "baz")"#,
```

Comment:
I find this _somewhat_ difficult to read. Would it be possible to do something like:

```coffee
insert!({"foo": { "bar": [] }}, "foo.bar[1]", "baz")
```

Reply:
Yeah, agreed, I Would have expected to be able to pass a path as the second argument rather than an array of path parts.

Reply:
Or, rather, I see it supports both here :) Maybe the examples could be using path strings primarily? I think that would be more common.

Reply:
I thought about it, but decided against it because:

1.  It leads to all sort of escaping-hell when you actually _want_ a key named `foo.bar[1]`.
2. Additionally, using an array of strings and integers means we (at some point) can make this function infallible, whereas we _can't_ do that if we accept a string that we then have to parse first to check for validity.
3. And finally, by using an array, it's way easier to build up paths from loose segments (which is mostly why this function exists), as opposed to having to write `my_foo + "." + my_bar + "[" + my_index + "]"`.

Your example looks cleaner when used as such, an example. But your example works even better if you'd just write `{"foo": { "bar": [] }}.foo.bar[1]`. Which works currently, but doesn't solve the need to construct paths dynamically.

Reply:
Yeah, I see where an array would be more readable. It seems like string paths are supported too though?

https://github.com/vectordotdev/vector/pull/9284/files#diff-0657e28e6e8b599377806d31db7d981cc754f51c8befdb79e29ea9e127efd202R98

Reply:
Well. Not exactly.

The string is when you just want to reference _one_ path. It's the "easy mode" of this function. The array is used when you want to go to a nested path. There should (is?) already be an example demonstrating that.

Wr.t. `LookupBuf`, the `From<String>` impl takes the literal string as a single path segment. The `from_str` is the one that parses a string-based-path such as `.foo.bar[2]` and can return an error if it's an invalid format. I'm using the `from` one, so you can't do nested path access with a string, you need the array for that.

Reply:
☝️ that could be confusing, given that - perhaps erring to only do ARRAY for clarity

Reply:
I don't think it's confusing, since we don't allow string-based `".foo.bar"` paths anywhere in VRL.

In fact, if you do `.foo."bar.baz"`, it works exactly as this function, the `bar.baz` part is treated as a single segment.

Reply:
> my gut just said `"foo.bar"` is a two segment path, rather than a single segment.

I hear you. I still don't think it's a good idea because of the before-mentioned three reasons, and the added fourth reason that we don't use that form of writing string-based paths anywhere in VRL.

Reply:
I'm fairly confident people are going to expect:

```
. = {"foo": {"bar": 5}}
lookup(., "foo.bar")
```

To return 5. I realize escaping is tricky if they do want `.`s in their path though. I think they could use the array-based access you added here to do that by doing `lookup(., ["foo.bar"])`.

Alternatively, we drop the string based path access and _only_ allow arrays. I'm just concerned users will expect that the paths are resolved including `.` based splitting.

Reply:
I'm _still_ leaning towards not allowing that. Even if there _might_ be people expecting that, we can (1) document it, and (2) we get the nice benefit of not having to parse the string path at runtime, which means we get better performance, but more importantly, the function can become infallible.

I mean, if it turns out to really be an issue, we can always change the default behaviour (as a breaking change).

I'm also happy to drop strings for now, and only allow arrays, to avoid any ambiguity, and to allow either direction to be taken in the future without any breaking changes.

Reply:
`lookup(., "foo.bar")` is largely analogous to doing `.["foo.bar"]`, so I think that the way it currently works is correct.

There would also be a slight performance impact to parsing the path each time.

Also the use case for this would not be passing a static string here it will be a variable - `lookup(., .field)`, which implies to me it is just a single field rather than a whole path to parse.

Reply:
I think the examples could be made easier to read if they specified the parameter names:

```
insert!({"foo": { "bar": [] }}, field: ["foo", "bar", 1], value: "baz")
```

Reply:
> `lookup(., "foo.bar")` is largely analogous to doing `.["foo.bar"]`, so I think that the way it currently works is correct.

I think this is a good point.

If we ever want to support dynamic lookup using dedicated syntax, then the one you wrote is by far the most likely candidate to be used. You would write `.["foo"]["bar"][1]` in such a case, not `.["foo.bar[1]"]`.

As stated before, I currently see two viable paths forward:

1. We use the implementation as-is, and support strings `"foo"` for non-nested dynamic field access, and arrays `["foo", "bar"]` for nested access.
2. We remove string-based dynamic field access entirely, and only support arrays, making non-nested dynamic field access a _bit_ less ergonomic (`["foo"]` vs `"foo"`), but keep our options open to which direction we want to go with direct string dynamic access.

I favour (1), but wouldn't mind (2) if people feel strongly about it.

Reply:
I'm in the (1) camp.

Reply:
My preference would be for 2 here, that we start with just arrays for paths. This:

* should make it clearer that the path is a list of parts
* leaves the door open to dynamic string path resolution later if users ask for it
* will remove any confusion about simple strings being treated as "flat". I realize we could document this behavior, but I think, as much as we can, having it do what users will expect is preferable. I'm fairly confident people will be confused by it being treated as flat.

Reply:
Alright, since not everyone is in the (1) camp, and given that (2) still allows for (1) to be realized if we so agree, I'm happy to not support strings for now.